src/viam/resource/types.py

```diff
@@ -41,4 +51,130 @@ def __eq__(self, other: "Subtype") -> bool:
+
+    family: str
+    """The family name"""
+
+    DEFAULT_FAMILY_NAME: ClassVar[str] = "builtin"
+
+    DEFAULT: ClassVar["ModelFamily"]
+
+    def __init__(self, namespace: str, family: str):
+        self.namespace = namespace.lower()
+        self.family = family.lower()
+
+    def __str__(self) -> str:
+        return f"{self.namespace}:{self.family}"
+
+    def __repr__(self) -> str:
+        return f"<viam.resource.types.ModelFamily {str(self)} at {hex(id(self))}>"
+
+    def __hash__(self) -> int:
+        return hash(str(self))
+
+    def __eq__(self, other: "ModelFamily") -> bool:
+        return str(self) == str(other)
+
+
+ModelFamily.DEFAULT = ModelFamily(RESOURCE_NAMESPACE_RDK, ModelFamily.DEFAULT_FAMILY_NAME)
+
+
+class Model:
+    """Represents a specific model within a family of models"""
+
+    model_family: ModelFamily
+    """The family to which this model belongs"""
+
+    name: str
+    """The name of the model"""
+
+    def __init__(self, model_family: ModelFamily, name: str):
+        self.model_family = model_family
+        self.name = name.lower()
+
+    def __str__(self) -> str:
+        return f"{self.model_family}:{self.name}"
+
+    def __repr__(self) -> str:
+        return f"<viam.resource.types.Model {str(self)} at {hex(id(self))}>"
+
+    def __hash__(self) -> int:
+        return hash(str(self))
+
+    def __eq__(self, other: "Model") -> bool:
+        return str(self) == str(other)
+
+    @classmethod
+    def from_string(cls, model: str, *, ignore_errors=False) -> Self:
+        """Create a ```Model``` from its string representation (namespace:family:name).
+
+        Args:
+            model (str): The Model as a string
+            ignore_errors (bool, optional): If namespace or family are not found in the string, default to empty string
+                                            rather than raise an exception. Defaults to False.
+
+        Raises:
+            ValueError: Raised if the provided string is not a valid Model
+
+        Returns:
+            Self: The Model
+        """
+        regex = re.compile(r"^([\w-]+):([\w-]+):([\w-]+)$")
+        match = regex.match(model)
+        if match:
+            namespace = match.group(1)
+            family = match.group(2)
+            name = match.group(3)
+            model_family = ModelFamily(namespace, family)
+        elif ignore_errors:
+            model_family = ModelFamily("", "")
+            name = model
+        else:
+            raise ValueError(f"{model} is not a valid Model")
+
+        return cls(model_family, name)
+
+
+def resource_name_from_string(string: str) -> ResourceName:
+    parts = string.split(":")
+    if len(parts) < 4:
+        raise ValueError(f"{string} is not a valid ResourceName")
+    return ResourceName(namespace=parts[0], type=parts[1], subtype=parts[2], name=":".join(parts[3:]))
+
+
+ComponentCreator: TypeAlias = Callable[[Mapping[ResourceName, "ComponentBase"], ComponentConfig], "ComponentBase"]
```

Comment:
Can we name this ResourceCreator to limit the amount of future changes we have to make?

Reply:
The reason I called it `ComponentCreator` for now was because I haven't yet determined what changes I'll ned to make to convert this to an actual `ResourceCreator` (specifically since the input and return types are still `ComponentBase` and not a nonexistent `ResourceBase`)

It's kind of like a breadcrumb/clue in the code that it needs more thought and updating

Reply:
Also because this type is used internally (users don't ever have to name this type specifically, just conform to it), renaming this later shouldn't be breaking